Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean

```diff
@@ -46,6 +82,204 @@ noncomputable def preservesFiniteColimitsOfPreservesHomology
+    F.PreservesEpimorphisms where
+  preserves f := h _ { exact := ShortComplex.exact_kernel f } |>.2
+
+/--
+For an addivite functor `F : C ⥤ D` between abelian categories, the following are equivalent:
+- `F` preserves short exact sequences on the right hand side, i.e. if `0 ⟶ A ⟶ B ⟶ C ⟶ 0` is
+  exact then `F(A) ⟶ F(B) ⟶ F(C) ⟶ 0` is exact.
+- `F` preserves exact sequences on the right hand side, i.e. if `A ⟶ B ⟶ C` is exact where `B ⟶ C`
+  is epi, then `F(A) ⟶ F(B) ⟶ F(C) ⟶ 0` is exact and `F(B) ⟶ F(C)` is epi as well.
+- `F` preserves cokernels.
+- `F` preserves finite colimits.
+-/
+lemma preservesFiniteColimits_tfae : List.TFAE
+    [
+      ∀ (S : ShortComplex C), S.ShortExact → (S.map F).Exact ∧ Epi (F.map S.g),
+      ∀ (S : ShortComplex C), S.Exact ∧ Epi S.g → (S.map F).Exact ∧ Epi (F.map S.g),
+      ∀ ⦃X Y : C⦄ (f : X ⟶ Y), Nonempty $ PreservesColimit (parallelPair f 0) F,
+      Nonempty $ PreservesFiniteColimits F
+    ] := by
+  tfae_have 1 → 2
+  · rintro hF S ⟨hS, hf⟩
+    have := preservesEpimorphism_of_preserves_shortExact_right F hF
+    refine ⟨?_, inferInstance⟩
+    let T := ShortComplex.mk (Abelian.image.ι S.f) S.g (Abelian.image_ι_comp_eq_zero S.zero)
+    let φ : S.map F ⟶ T.map F :=
+      { τ₁ := F.map $ (Abelian.factorThruImage S.f)
+        τ₂ := 𝟙 _
+        τ₃ := 𝟙 _
+        comm₁₂ := by
+          dsimp
+          rw [← F.map_comp, Abelian.image.fac, Category.comp_id] }
+    exact (ShortComplex.exact_iff_of_epi_of_isIso_of_mono φ).2
+      (hF T { exact := (S.exact_iff_exact_image_ι).1 hS }).1
+
+  tfae_have 2 → 3
+  · intro hF X Y f
+    refine ⟨preservesColimitOfPreservesColimitCocone (cokernelIsCokernel f) ?_⟩
+    apply (CokernelCofork.isColimitMapCoconeEquiv _ F).2
+    let S := ShortComplex.mk _ _ (cokernel.condition f)
+    let hS := hF S ⟨ShortComplex.exact_cokernel f, by infer_instance⟩
+    have : Epi (S.map F).g := hS.2
+    exact hS.1.gIsCokernel
+
+  tfae_have 3 → 4
+  · intro hF
+    have := fun X Y (f : X ⟶ Y) ↦ (hF f).some
+    exact ⟨by apply preservesFiniteColimitsOfPreservesCokernels⟩
+
+  tfae_have 4 → 1
+  · rintro ⟨_⟩ S hS
+    exact (S.map F).exact_and_epi_g_iff_g_is_cokernel |>.2
+      ⟨CokernelCofork.mapIsColimit _ hS.gIsCokernel F⟩
+
+  tfae_finish
+
+open ZeroObject in
+/--
+For an additive functor `F : C ⥤ D` between abelian categories, the following are equivalent:
+- `F` preserves short exact sequences, i.e. if `0 ⟶ A ⟶ B ⟶ C ⟶ 0` is exact then
+  `0 ⟶ F(A) ⟶ F(B) ⟶ F(C) ⟶ 0` is exact.
+- `F` preserves exact sequences, i.e. if `A ⟶ B ⟶ C` is exact then `F(A) ⟶ F(B) ⟶ F(C)` is exact.
+- `F` preserves both finite limits and finite colimits.
+- `F` preserves homology.
+-/
+lemma exact_tfae : List.TFAE
+    [
+      ∀ (S : ShortComplex C), S.ShortExact → (S.map F).ShortExact,
+      ∀ (S : ShortComplex C), S.Exact → (S.map F).Exact,
+      Nonempty (PreservesFiniteLimits F) ∧ Nonempty (PreservesFiniteColimits F),
+      Nonempty (PreservesHomology F)
```

Comment:
The statement would be more parallel to the previous lemmas if 3 and 4 were exchanged.